File: leetcode/linked-list/206. Reverse Linked List/py/main.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next


# Time O(N)
# Space O(N)
class Solution:
    def reverseList(self, node):
        head = None
        current = node
        
        while current:
            new_node = ListNode(current.val)
            new_node.next = head
            
            head = new_node
            current = current.next

        return head

# Time O(N)
# Space O(N)
class Solution_II:
    def reverseList(self, head):  
        def dfs(head):
            if head == None:
                return None
            
            if head.next == None:
                return head
            
            prev = dfs(head.next)
            
            head.next.next = head
            head.next = None
            
            return prev
        
        return dfs(head)

File: leetcode/linked-list/206. Reverse Linked List/js/main.js
// Time O(N)
// Space O(N)
const reverseList = head => {
  if (head === null) return null;
  if (head.next === null) return head;

  let dummy = reverseList(head.next);

  head.next.next = head;
  head.next = null;

  return dummy;
};

/*
  Пример: 

  1 -> 2 -> 3
  
  cur - 1
  prev - null
  ------------
  cur - 2
  prev - 1
  ------------
  cur - 3
  prev - 2 -> 1
  ------------
  cur - null
  prev - 3 -> 2 -> 1

*/

// Time complexity : O(N). где N длинна листа
// Space complexity : O(1).
const reverseList_II = root => {
  // prev будет содержать head reverse node
  let prev = null;
  let curr = root;

  while (curr !== null) {
    let nextTemp = curr.next;

    curr.next = prev;
    prev = curr;
    curr = nextTemp;
  }

  return prev;
};
